File: backend/src/main/kotlin/com/msp/doku/service/DocumentationService.kt
package com.msp.doku.service

import com.msp.doku.domain.DocumentationSection
import com.msp.doku.dto.DocumentationOverviewDto
import com.msp.doku.dto.DocumentationSectionDto
import com.msp.doku.dto.UpdateDocumentationRequest
import com.msp.doku.repository.DocumentationRepository
import com.msp.doku.repository.TenantRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class DocumentationService(
    private val documentationRepository: DocumentationRepository,
    private val tenantRepository: TenantRepository
) {

    companion object {
        val SECTION_TYPES = listOf(
            "access_credentials",
            "network_architecture",
            "critical_services",
            "backup_recovery",
            "monitoring_alerting",
            "sops",
            "change_update",
            "security_baseline",
            "disaster_recovery",
            "external_integrations",
            "naming_conventions"
        )

        val SECTION_SCHEMAS: Map<String, List<String>> = mapOf(
            "access_credentials" to listOf("mfaRequirements", "breakGlassAccounts", "privilegedAccessPaths", "adminAccessConcept"),
            "network_architecture" to listOf("vlanStructure", "routingDesign", "internetBreakout", "siteToSiteVpn", "wanSetup"),
            "critical_services" to listOf("businessCriticalSystems", "internalDependencies", "externalDependencies"),
            "backup_recovery" to listOf("backupTargets", "retentionPolicy", "rtoRpo", "recoveryProcedures"),
            "monitoring_alerting" to listOf("monitoredSystems", "alertThresholds", "alertDestinations", "escalationFlow"),
            "sops" to listOf("tenantOnboarding", "serverDeployment", "vmRestore", "incidentHandling"),
            "change_update" to listOf("patchManagement", "maintenanceWindows", "changeApproval"),
            "security_baseline" to listOf("hardeningStandards", "avEdrStrategy", "loggingStrategy", "accessControlModel"),
            "disaster_recovery" to listOf("datacenterDown", "ransomwareResponse", "fullTenantLoss", "recoveryOrder"),
            "external_integrations" to listOf("isps", "cloudProviders", "saasTools", "licensingDependencies"),
            "naming_conventions" to listOf("serverNaming", "networkNaming", "vlanNaming", "generalConventions")
        )
    }

    fun getOverview(tenantId: UUID): List<DocumentationOverviewDto> {
        val existing = documentationRepository.findByTenantId(tenantId)
            .associateBy { it.sectionType }

        return SECTION_TYPES.map { type ->
            val section = existing[type]
            val schema = SECTION_SCHEMAS[type] ?: emptyList()
            val filledFields = section?.structuredData?.count { it.value.toString().isNotBlank() } ?: 0
            val totalFields = schema.size.coerceAtLeast(1)

            DocumentationOverviewDto(
                sectionType = type,
                exists = section != null,
                updatedAt = section?.updatedAt,
                completionPercent = if (section != null) ((filledFields.toDouble() / totalFields) * 100).toInt() else 0
            )
        }
    }

    fun getSection(tenantId: UUID, sectionType: String): DocumentationSectionDto {
        val section = documentationRepository.findByTenantIdAndSectionType(tenantId, sectionType)
        if (section != null) return section.toDto()

        // Return empty section with schema fields
        val schema = SECTION_SCHEMAS[sectionType] ?: emptyList()
        return DocumentationSectionDto(
            id = UUID.randomUUID(),
            sectionType = sectionType,
            structuredData = schema.associateWith { "" },
            notes = null,
            updatedBy = null,
            version = 0,
            createdAt = null,
            updatedAt = null
        )
    }

    @Transactional
    fun updateSection(tenantId: UUID, sectionType: String, request: UpdateDocumentationRequest): DocumentationSectionDto {
        val tenant = tenantRepository.findById(tenantId)
            .orElseThrow { IllegalArgumentException("Tenant not found") }

        var section = documentationRepository.findByTenantIdAndSectionType(tenantId, sectionType)

        if (section == null) {
            section = DocumentationSection(
                tenant = tenant,
                sectionType = sectionType,
                structuredData = request.structuredData ?: emptyMap(),
                notes = request.notes,
                version = 1
            )
        } else {
            request.structuredData?.let { section.structuredData = it }
            request.notes?.let { section.notes = it }
            section.version += 1
        }

        return documentationRepository.save(section).toDto()
    }

    private fun DocumentationSection.toDto() = DocumentationSectionDto(
        id = this.id!!,
        sectionType = this.sectionType,
        structuredData = this.structuredData,
        notes = this.notes,
        updatedBy = this.updatedBy,
        version = this.version,
        createdAt = this.createdAt,
        updatedAt = this.updatedAt
    )
}

File: frontend/src/services/DocumentationService.ts
import { apiFetch } from './apiClient';

export interface DocumentationOverview {
    sectionType: string;
    exists: boolean;
    updatedAt: string | null;
    completionPercent: number;
}

export interface DocumentationSection {
    id: string;
    sectionType: string;
    structuredData: Record<string, any>;
    notes: string | null;
    updatedBy: string | null;
    version: number;
    createdAt: string | null;
    updatedAt: string | null;
}

export interface UpdateDocumentationRequest {
    structuredData?: Record<string, any>;
    notes?: string;
}

export const SECTION_LABELS: Record<string, string> = {
    access_credentials: 'Access & Credentials',
    network_architecture: 'Network Architecture',
    critical_services: 'Critical Services & Dependencies',
    backup_recovery: 'Backup & Recovery',
    monitoring_alerting: 'Monitoring & Alerting',
    sops: 'Standard Operating Procedures',
    change_update: 'Change & Update Strategy',
    security_baseline: 'Security Baseline',
    disaster_recovery: 'Disaster Recovery',
    external_integrations: 'External Integrations',
    naming_conventions: 'Naming Conventions',
};

export const FIELD_LABELS: Record<string, Record<string, string>> = {
    access_credentials: {
        mfaRequirements: 'MFA Setup Requirements',
        breakGlassAccounts: 'Break-Glass Accounts',
        privilegedAccessPaths: 'Privileged Access Paths (VPN → Jump Host → Systems)',
        adminAccessConcept: 'Admin Access Concept',
    },
    network_architecture: {
        vlanStructure: 'VLAN Structure & Purpose',
        routingDesign: 'Routing Design (Core Switch / Firewall Roles)',
        internetBreakout: 'Internet Breakout / WAN Setup',
        siteToSiteVpn: 'Site-to-Site VPN Overview',
        wanSetup: 'WAN Setup',
    },
    critical_services: {
        businessCriticalSystems: 'Business-Critical Systems',
        internalDependencies: 'Internal Dependencies (AD → DNS → DHCP → Apps)',
        externalDependencies: 'External Dependencies (Cloud, SaaS, APIs)',
    },
    backup_recovery: {
        backupTargets: 'What Is Backed Up',
        retentionPolicy: 'Retention Policies',
        rtoRpo: 'RTO / RPO Expectations',
        recoveryProcedures: 'Recovery Procedures (Step-by-Step)',
    },
    monitoring_alerting: {
        monitoredSystems: 'What Is Monitored',
        alertThresholds: 'Alert Thresholds',
        alertDestinations: 'Where Alerts Go (Mail, Teams, etc.)',
        escalationFlow: 'Escalation Flow',
    },
    sops: {
        tenantOnboarding: 'New Customer / Tenant Onboarding',
        serverDeployment: 'Server Deployment Standard',
        vmRestore: 'VM Restore Process',
        incidentHandling: 'Incident Handling Steps',
    },
    change_update: {
        patchManagement: 'Patch Management Approach',
        maintenanceWindows: 'Maintenance Windows',
        changeApproval: 'Change Approval Logic',
    },
    security_baseline: {
        hardeningStandards: 'Hardening Standards (CIS-Level Summary)',
        avEdrStrategy: 'AV/EDR Strategy',
        loggingStrategy: 'Logging (What + Where)',
        accessControlModel: 'Access Control Model (RBAC Concept)',
    },
    disaster_recovery: {
        datacenterDown: 'Scenario: Datacenter Down',
        ransomwareResponse: 'Scenario: Ransomware',
        fullTenantLoss: 'Scenario: Full Tenant Loss',
        recoveryOrder: 'Recovery Order (What Comes First)',
    },
    external_integrations: {
        isps: 'ISPs',
        cloudProviders: 'Cloud Providers',
        saasTools: 'SaaS Tools',
        licensingDependencies: 'Licensing Dependencies',
    },
    naming_conventions: {
        serverNaming: 'Server Naming',
        networkNaming: 'Network Device Naming',
        vlanNaming: 'VLAN Naming',
        generalConventions: 'General Conventions',
    },
};

export const DocumentationService = {
    getOverview: (tenantId: string) =>
        apiFetch<DocumentationOverview[]>(`/tenants/${tenantId}/docs`),
    getSection: (tenantId: string, sectionType: string) =>
        apiFetch<DocumentationSection>(`/tenants/${tenantId}/docs/${sectionType}`),
    updateSection: (tenantId: string, sectionType: string, data: UpdateDocumentationRequest) =>
        apiFetch<DocumentationSection>(`/tenants/${tenantId}/docs/${sectionType}`, {
            method: 'PUT',
            body: JSON.stringify(data),
        }),
};
